chain/chain/src/chain.rs

```diff
@@ -1968,6 +1970,160 @@ impl Chain {
+    fn get_node(
+        &mut self,
+        index: u64,
+        level: u64,
+        counter: u64,
+        tree_size: u64,
+        tree_nodes: &mut HashMap<(u64, u64), MerkleHash>,
+    ) -> Result<Option<MerkleHash>, Error> {
+        if let Some(hash) = tree_nodes.get(&(index, level)) {
+            Ok(Some(*hash))
+        } else {
+            if level == 0 {
+                if index >= tree_size {
+                    return Ok(None);
+                }
+                let hash = *self.mut_store().get_block_hash_from_ordinal(index)?;
+                tree_nodes.insert((index, level), hash);
+                Ok(Some(hash))
+            } else {
+                let cur_tree_size = (index + 1) * counter;
+                if cur_tree_size > tree_size {
+                    let left_hash =
+                        self.get_node(index * 2, level - 1, counter / 2, tree_size, tree_nodes)?;
+                    let right_hash = self.reconstruct_node(
+                        index * 2 + 1,
+                        level - 1,
+                        counter / 2,
+                        tree_size,
+                        tree_nodes,
+                    )?;
+                    return Ok(Self::combine_maybe_hashes(left_hash, right_hash));
+                }
+                let hash = *self
+                    .mut_store()
+                    .get_block_merkle_tree_from_ordinal(cur_tree_size)?
+                    .get_path()
+                    .last()
+                    .ok_or_else(|| ErrorKind::Other("Merkle tree node missing".to_string()))?;
+                tree_nodes.insert((index, level), hash);
+                Ok(Some(hash))
+            }
+        }
+    }
+
+    /// Reconstruct node at given position (index, level). If the node does not exist, return `None`.
+    fn reconstruct_node(
+        &mut self,
+        index: u64,
+        level: u64,
+        counter: u64,
+        tree_size: u64,
+        tree_nodes: &mut HashMap<(u64, u64), MerkleHash>,
+    ) -> Result<Option<MerkleHash>, Error> {
+        if let Some(hash) = tree_nodes.get(&(index, level)) {
+            Ok(Some(*hash))
+        } else {
+            if level == 0 {
+                if index >= tree_size {
+                    return Ok(None);
+                }
+                let hash = *self.mut_store().get_block_hash_from_ordinal(index)?;
+                tree_nodes.insert((index, level), hash);
+                Ok(Some(hash))
+            } else {
+                let left_hash =
+                    self.get_node(index * 2, level - 1, counter / 2, tree_size, tree_nodes)?;
+                let right_hash = self.reconstruct_node(
+                    index * 2 + 1,
+                    level - 1,
+                    counter / 2,
+                    tree_size,
+                    tree_nodes,
+                )?;
+                let hash = Self::combine_maybe_hashes(left_hash, right_hash);
+                if let Some(h) = hash {
```

Comment:
Why only memoizing `Some`s, but not `None`s?

Reply:
sheer stupidity.